source/src_lcao/wfc_dm_2d.cpp

```diff
@@ -28,117 +30,190 @@ void Wfc_Dm_2d::init()
 	}
 }
 
-void Wfc_Dm_2d::cal_dm(const ModuleBase::matrix &wg)
+
+//for gamma_only
+void Wfc_Dm_2d::cal_dm(const ModuleBase::matrix& wg,
+    std::vector<ModuleBase::matrix>& wfc_gamma,
+    std::vector<ModuleBase::matrix>& dm_gamma)
 {
 	ModuleBase::TITLE("Wfc_Dm_2d", "cal_dm");
 	
 	#ifdef TEST_DIAG
-	{
-		static int istep=0;
-		std::ofstream ofs("wfc_"+ModuleBase::GlobalFunc::TO_STRING(istep++)+"_"+ModuleBase::GlobalFunc::TO_STRING(GlobalV::MY_RANK));
-		if(GlobalV::GAMMA_ONLY_LOCAL)
-		{
-			ofs<<wfc_gamma<<std::endl;
-		}
-		else
-		{
-			ofs<<wfc_k<<std::endl;
-		}
-	}
+    static int istep=0;
+    std::ofstream ofs("wfc_"+ModuleBase::GlobalFunc::TO_STRING(istep++)+"_"+ModuleBase::GlobalFunc::TO_STRING(GlobalV::MY_RANK));
+    ofs<<wfc_gamma<<std::endl;
+	#endif
+	
+	// dm = wfc.T * wg * wfc.conj()
+	// dm[is](iw1,iw2) = \sum_{ib} wfc[is](ib,iw1).T * wg(is,ib) * wfc[is](ib,iw2).conj()
+	assert(wg.nc<=GlobalV::NLOCAL);
+    assert(wg.nr==GlobalV::NSPIN);
+    for(int is=0; is!=GlobalV::NSPIN; ++is)
+    {
+        std::vector<double> wg_local(GlobalC::ParaO.ncol,0.0);
+        for(int ib_global=0; ib_global!=wg.nc; ++ib_global)
+        {
+            const int ib_local = GlobalC::ParaO.trace_loc_col[ib_global];
+            if(ib_local>=0)
+            {
+                wg_local[ib_local] = wg(is,ib_global);
+            }
+        }
+        
+        // wg_wfc(ib,iw) = wg[ib] * wfc(ib,iw);
+        ModuleBase::matrix wg_wfc(wfc_gamma[is]);
+        for(int ir=0; ir!=wg_wfc.nr; ++ir)
+        {
+            BlasConnector::scal( wg_wfc.nc, wg_local[ir], wg_wfc.c+ir*wg_wfc.nc, 1 );
+        }
+
+        // C++: dm(iw1,iw2) = wfc(ib,iw1).T * wg_wfc(ib,iw2)
+        const double one_float=1.0, zero_float=0.0;
+        const int one_int=1;
+        const char N_char='N', T_char='T';
+        dm_gamma[is].create( wfc_gamma[is].nr, wfc_gamma[is].nc );
+        pdgemm_(
+            &N_char, &T_char,
+            &GlobalV::NLOCAL, &GlobalV::NLOCAL, &wg.nc,
+            &one_float,
+            wg_wfc.c, &one_int, &one_int, GlobalC::ParaO.desc,
+            wfc_gamma[is].c, &one_int, &one_int, GlobalC::ParaO.desc,
+            &zero_float,
+            dm_gamma[is].c, &one_int, &one_int, GlobalC::ParaO.desc);
+    }
+    #ifdef TEST_DIAG
+    static int istep=0;
+    std::ofstream ofs("dm_" + ModuleBase::GlobalFunc::TO_STRING(istep) + "_" + ModuleBase::GlobalFunc::TO_STRING(GlobalV::MY_RANK));
+    ofs << dm_gamma << std::endl;
+	#endif
+
+	return;
+}
+
+//for multi-k
+void Wfc_Dm_2d::cal_dm(const ModuleBase::matrix& wg,    // wg(ik,ib), cal all dm 
```

Comment:
This function looks similar with the function above, why not use template to combine these two functions?